File: December 2022/19 December 2022/Java/Alex Travelling.java
//{ Driver Code Starts
// Initial Template for Java

import java.util.*;
import java.lang.*;
import java.io.*;

// Position this line where user code will be pasted.

class GFG {
    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        int T = sc.nextInt();
        while (T-- > 0) {
            int n = sc.nextInt();
            int k = sc.nextInt();

            int tmp = sc.nextInt();
            int[][] flights = new int[tmp][3];
            for (int i = 0; i < tmp; i++) {

                int u1 = sc.nextInt();
                int v1 = sc.nextInt();
                int w1 = sc.nextInt();
                flights[i][0] = u1;
                flights[i][1] = v1;
                flights[i][2] = w1;
            }

            Solution ob = new Solution();
            int ans = ob.minimumCost(flights, n, k);
            System.out.println(ans);
        }
    }
}
// } Driver Code Ends

class Solution {

    int[] dist;
    class Node {

        int v;
        int w;

        Node(int v, int w) {
            this.v = v;
            this.w = w;
        }
    }

    int minimumCost(int[][] flights, int n, int k) {
        List<List<Node>> adj = new ArrayList<>();

        for (int i = 0; i <= n; i++) {
            adj.add(new ArrayList<>());
        }

        for (int[] i : flights) {
            adj.get(i[0]).add(new Node(i[1], i[2]));
        }
        dist = new int[n + 1];
        Arrays.fill(dist, 100000);
        dijkistra(adj, k);
        int max = 0;
        for (int i = 1; i <= n; i++) {
            max = Math.max(max, dist[i]);
        }

        return max == 100000 ? -1 : max;
    }

    public void dijkistra(List<List<Node>> adj, int start) {

        PriorityQueue<Node> pq =
            new PriorityQueue<>((x, y)->{ return x.w - y.w; });

        dist[start] = 0;

        pq.add(new Node(start, 0));

        while (!pq.isEmpty()) {
            Node vet = pq.poll();

            for (Node child : adj.get(vet.v)) {

                if (dist[vet.v] + child.w < dist[child.v]) {
                    dist[child.v] = dist[vet.v] + child.w;
                    pq.add(new Node(child.v, dist[child.v]));
                }
            }
        }
    }
}

File: December 2022/19 December 2022/C++/Alex Travelling.cpp
//{ Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends

class Solution {
  public:
    vector<pair<int, int>> adj[501];

    void dijkstra(vector<int>& costToReach, int source, int n) {
        priority_queue<pair<int, int>, vector<pair<int, int>>,
                       greater<pair<int, int>>> pq;
        pq.push({0, source});

        costToReach[source] = 0;

        while (!pq.empty()) {
            int currNodeCost = pq.top().first;
            int currNode = pq.top().second;
            pq.pop();

            if (currNodeCost > costToReach[currNode]) {
                continue;
            }

            for (pair<int, int> edge : adj[currNode]) {
                int cost = edge.first;
                int neighborNode = edge.second;

                if (costToReach[neighborNode] > currNodeCost + cost) {
                    costToReach[neighborNode] = currNodeCost + cost;
                    pq.push({costToReach[neighborNode], neighborNode});
                }
            }
        }
    }
    int minimumCost(vector<vector<int>>& flights, int n, int k) {
        for (vector<int> f : flights) {
            int source = f[0];
            int dest = f[1];
            int travelCost = f[2];

            adj[source].push_back({travelCost, dest});
        }

        vector<int> costToReach(n + 1, INT_MAX);
        dijkstra(costToReach, k, n);

        int answer = INT_MIN;
        for (int i = 1; i <= n; i++) {
            answer = max(answer, costToReach[i]);
        }

        return answer == INT_MAX ? -1 : answer;
    }
};


//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    while (t--) {
        int n, k;
        cin >> n >> k;
        int size;
        cin >> size;
        vector<vector<int>> flights(size,vector<int>(3));
        for (int i = 0; i < size; i++) {
            vector<int> temp;
            for (int j = 0; j < 3; ++j) {
                int x;
                cin >> x;
                temp.push_back(x);
            }
            flights[i]=temp;
        }

        Solution ob;
        cout << ob.minimumCost(flights, n, k) << "\n";
    }
}

// } Driver Code Ends